### PR TITLE
lint tests with mypy in CI and pre-commit

### PR DESCRIPTION
**Pull request by roelofvandijk**
* I added a host of per-file exclusions. This means all new files are linted, all files that are not excluded (those that pass now) keep being linted, and we can start fixing the typing errors file-by-file.
~* I also included some minor fixes - those can be moved to a separate PR if preferred.~
* `strict-equality` can move to the config file, and the included files as well.

Pre-commit and CI now just call `mypy`
Running pre-commit in the ci would make this even easier to maintain.

Here is the number of remaining issues per file:
```
 241 test/external/external_test_hcq.py
 189 test/test_hcq.py
 113 extra/onnx_ops.py
  97 examples/so_vits_svc.py
  61 test/test_nn.py
  57 test/test_ops.py
  47 extra/backends/cstyle.py
  42 test/test_tensor.py
  36 extra/archprobe.py
  32 extra/onnx.py
  31 examples/vits.py
  31 examples/hlb_cifar10.py
  30 test/external/external_test_nv.py
  28 test/imported/test_indexing.py
  27 extra/datasets/openimages.py
  26 test/test_fuzz_shape_ops.py
  26 extra/thneed.py
  26 extra/backends/ops_hip.py
  26 examples/whisper.py
  25 extra/models/mask_rcnn.py
  23 examples/conversation.py
  22 test/unit/test_shapetracker.py
  22 test/external/speed_compare_cuda_ptx.py
  22 examples/yolov3.py
  20 test/external/external_test_amd.py
  20 test/external/external_benchmark_resnet.py
  19 test/test_linearizer.py
  19 examples/stable_diffusion.py
  18 test/test_image_dtype.py
  18 extra/hip_gpu_driver/test_pm4.py
  17 extra/datasets/wikipedia.py
  17 extra/assembly/assembly.py
  16 test/models/test_whisper.py
  16 test/models/test_efficientnet.py
  16 test/external/speed_compare_cuda_nv.py
  16 extra/qcom_gpu_driver/msm_kgsl.py
  16 extra/export_model.py
  16 examples/yolov8.py
  16 examples/mlperf/model_train.py
  14 test/external/mlperf_bert/external_benchmark_bert.py
  14 test/external/external_test_opt.py
  14 examples/vit.py
  14 examples/mlperf/dataloader.py
  13 test/test_net_speed.py
  13 test/test_assign.py
  13 extra/models/retinanet.py
  13 extra/disk_read_speed.py
  13 extra/assembly/rocm/rdna3/asm.py
  13 examples/openpilot/compile2.py
  13 examples/llama3.py
  12 extra/optimization/helpers.py
  12 examples/sovits_helpers/preprocess.py
  11 test/unit/test_disk_tensor.py
  11 test/test_uops.py
  11 test/models/test_end2end.py
  11 extra/models/unet.py
  11 examples/vgg7.py
  11 examples/sdxl.py
  10 extra/models/clip.py
  10 extra/backends/triton.py
   9 test/test_device_speed.py
   9 test/models/test_real_world.py
   9 extra/backends/ops_webgpu.py
   9 extra/assembly/assembly_ptx.py
   9 examples/transformer.py
   8 test/test_schedule.py
   8 extra/optimization/extract_sa_pairs.py
   8 extra/gemm/metal_matvec.py
   8 extra/assembly/assembly_rdna.py
   8 extra/assembly/assembly_arm64.py
   8 extra/accel/ane/lib/ane.py
   8 extra/accel/ane/2_compile/ane.py
   8 examples/mamba.py
   8 examples/llm.c/export.py
   7 test/unit/test_tqdm.py
   7 test/test_speed_v_torch.py
   7 test/test_pattern_matcher.py
   7 test/test_optim.py
   7 test/external/external_test_uops_graphing.py
   7 test/external/external_osx_profiling.py
   7 extra/datasets/kits19.py
   7 examples/webgpu/stable_diffusion/compile.py
   7 examples/handcode_opt.py
   6 test/external/fuzz_kfd.py
   6 test/external/external_test_datasets.py
   6 test/external/external_benchmark_multitensor_allreduce.py
   6 extra/backends/ops_webgl.py
   6 examples/train_efficientnet.py
   6 examples/llm.c/train_gpt2.py
   6 examples/coder.py
   5 test/test_fusion_op.py
   5 test/test_dtype_alu.py
   5 test/external/fuzz_graph.py
   5 test/external/external_test_yolo.py
   5 test/external/external_test_jit_on_models.py
   5 extra/lr_scheduler.py
   5 examples/serious_mnist.py
   5 docs/abstractions2.py
   4 test/test_symbolic_shapetracker.py
   4 test/test_multitensor.py
   4 test/test_method_cache.py
   4 test/external/fuzz_linearizer.py
   4 extra/models/rnnt.py
   4 extra/models/resnet.py
   4 extra/models/efficientnet.py
   4 extra/gemm/metal_matmul.py
   4 extra/gemm/intel_xmx.py
   4 extra/gemm/cuda_matmul.py
   4 extra/backends/rdna.py
   4 examples/gpt2.py
   4 examples/compile_tensorflow.py
   3 test/unit/test_helpers.py
   3 test/test_lazybuffer.py
   3 test/test_graph.py
   3 test/test_const_folding.py
   3 test/external/mlperf_bert/preprocessing/tokenization.py
   3 test/external/fuzz_symbolic.py
   3 test/external/fuzz_shapetracker.py
   3 test/external/external_test_yolov8.py
   3 test/external/external_test_example.py
   3 extra/optimization/rl.py
   3 extra/models/convnext.py
   3 extra/junk/sentencepiece_model_pb2.py
   3 extra/datasets/librispeech.py
   3 extra/datasets/coco.py
   3 extra/augment.py
   3 extra/assembly/ptx/test.py
   3 extra/accel/ane/2_compile/hwx_parse.py
   3 examples/mask_rcnn.py
   2 tinygrad/tensor.py
   2 test/test_zero_copy.py
   2 test/test_tensor_data.py
   2 test/test_search.py
   2 test/test_lazyop.py
   2 test/models/test_train.py
   2 test/external/mlperf_bert/preprocessing/create_pretraining_data.py
   2 test/external/mlperf_bert/external_test_checkpoint_loading.py
   2 test/external/external_test_speed_llama.py
   2 test/external/external_test_hip_compile.py
   2 test/external/external_benchmark_load_stable_diffusion.py
   2 test/external/external_benchmark_hip_compile.py
   2 sz.py
   2 extra/to_movement_ops.py
   2 extra/optimization/test_time_linearizer.py
   2 extra/models/llama.py
   2 extra/models/bert.py
   2 extra/gemm/tvm_gemm.py
   2 extra/gemm/metal_conv.py
   2 extra/gemm/hip_matmul.py
   2 extra/datasets/wikipedia_download.py
   2 extra/datasets/squad.py
   2 extra/datasets/imagenet.py
   2 extra/datasets/__init__.py
   2 extra/backends/graph_hip.py
   2 extra/accel/ane/amfi/new_patch.py
   2 examples/mlperf/initializers.py
   2 examples/mlperf/helpers.py
   2 examples/llama.py
   1 test/unit/test_disk_cache.py
   1 test/testextra/test_lr_scheduler.py
   1 test/test_winograd.py
   1 test/test_uops_stats.py
   1 test/test_setitem.py
   1 test/test_randomness.py
   1 test/test_pickle.py
   1 test/test_linearizer_failures.py
   1 test/test_kernel_cache.py
   1 test/test_jit.py
   1 test/models/test_mnist.py
   1 test/external/fuzz_shapetracker_math.py
   1 test/external/external_test_hsa_driver.py
   1 test/external/external_hip_compiler_bug.py
   1 test/external/external_cl_half_max.py
   1 extra/qcom_gpu_driver/opencl_ioctl.py
   1 extra/optimization/test_net.py
   1 extra/multitensor.py
   1 extra/hip_gpu_driver/test_sdma_fun.py
   1 extra/gradcheck.py
   1 extra/datasets/preprocess_imagenet.py
   1 extra/accel/ane/2_compile/struct_recover.py
   1 examples/webgl/yolov8/compile.py
   1 examples/rl/lightupbutton.py
   1 examples/openelm.py
   1 examples/mlperf/model_eval.py
   1 examples/beautiful_mnist_multigpu.py
```

### CODE REVIEW

File: test/test_linearizer.py [hunk 1]
```diff
@@ -1047,7 +1047,7 @@ def test_matvec(self):
     assert k.upcasted == 1
 
 def helper_linearizer_ast(_ast:Tuple[LazyOp, ...], inputs:List[Tensor], *args, **kwargs):
-  if not isinstance(_ast, LazyOp): ast = LazyOp(MetaOps.SINK, _ast)
+  ast = LazyOp(MetaOps.SINK, _ast)
```
Comment (roelofvandijk) [inline test/test_linearizer.py:1050]: Can never be LazyOp according to typing. I checked if it ever was, it wasn't.

File: test/test_linearizer.py [hunk 2]
```diff
@@ -209,7 +209,7 @@ def test_range_outer_op_after_phi_nested_range(self):
...
...s=False),))))), arg=None),), arg=None)), arg=None), LazyOp(op=BinaryOps.ADD, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=MemBuffer(idx=1, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(3, 27, 32), strides=(864, 32, 1), offset=0, mask=None, contiguous=True),)))), LazyOp(op=UnaryOps.NEG, src=(LazyOp(op=BinaryOps.MUL, src=(LazyOp(op=ReduceOps.SUM, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=MemBuffer(idx=1, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(3, 27, 32), strides=(864, 32, 1), offset=0, mask=None, contiguous=True),)))),), arg=(2,)), LazyOp(op=BufferOps.CONST, src=(), arg=ConstBuffer(val=0.03125, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(3, 27, 32), strides=(0, 0, 0), offset=0, mask=None, contiguous=False),))))), arg=None),), arg=None)), arg=None)), arg=None),), arg=(2,)), LazyOp(op=BufferOps.CONST, src=(), arg=ConstBuffer(val=0.03125, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(3, 27, 1), strides=(27, 1, 0), offset=0, mask=None, contiguous=True),))))), arg=None),), arg=None),), arg=MemBuffer(idx=0, dtype=dtypes.float, st=ShapeTracker(views=(View(shape=(3, 27, 1), strides=(27, 1, 0), offset=0, mask=None, contiguous=True),)))) # noqa: E501
```
Comment (roelofvandijk) [inline test/test_linearizer.py:212]: Sneaky comma at the end that makes this into a tuple moved into `helper_linearizer_ast`

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -15,7 +15,7 @@ repos:
         pass_filenames: false
       - id: mypy
         name: mypy
-        entry: mypy tinygrad/ --strict-equality
+        entry: mypy
```
Comment (roelofvandijkO) [inline .pre-commit-config.yaml:18]: Moved to `mypy.ini`

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -123,7 +123,7 @@ jobs:
     - name: Lint tinygrad with pylint
       run: python -m pylint tinygrad/
     - name: Run mypy
-      run: python -m mypy --strict-equality
+      run: python -m mypy
```
Comment (roelofvandijkO) [inline .github/workflows/test.yml:126]: Moved to `mypy.ini`

**Pull request closed without merging**